File: typescript/src/stack/generate-parentheses.ts
/**
 * LeetCode 22. Generate Parentheses (Medium)
 * https://leetcode.com/problems/generate-parentheses/
 *
 * Given n pairs of parentheses, generate all combinations of
 * well-formed parentheses.
 *
 * Run just this file:   npx tsx src/stack/generate-parentheses.ts
 * Run its tests:        npx vitest run src/stack/generate-parentheses.test.ts
 */

export function generateParenthesis(n: number): string[] {
  // TODO: implement
  throw new Error("Not implemented");
}

// Quick manual run / debugging playground — runs only when executed directly.
if (import.meta.url === `file://${process.argv[1]}`) {
  console.log(generateParenthesis(3)); // expected ["((()))","(()())","(())()","()(())","()()()"] in any order
}

File: python/stack/generate_parentheses.py
"""
LeetCode 22. Generate Parentheses (Medium)
https://leetcode.com/problems/generate-parentheses/

Given n pairs of parentheses, generate all combinations of
well-formed parentheses.

Run just this file:   python stack/generate_parentheses.py
Run its tests:        pytest stack/generate_parentheses.py -v
"""


class Solution:
    def generateParenthesis(self, n: int) -> list[str]:
        ...  # TODO: implement


# ---------------------------- tests ----------------------------
# Output order doesn't matter — compare sorted.

def test_example_1():
    assert sorted(Solution().generateParenthesis(3)) == sorted(
        ["((()))", "(()())", "(())()", "()(())", "()()()"]
    )


def test_example_2():
    assert sorted(Solution().generateParenthesis(1)) == ["()"]


def test_n_2():
    assert sorted(Solution().generateParenthesis(2)) == sorted(["(())", "()()"])


if __name__ == "__main__":
    # Debugging playground — set breakpoints here.
    s = Solution()
    print(s.generateParenthesis(3))  # expected ["((()))","(()())","(())()","()(())","()()()"] in any order
